Support new push notification payload format for multi-store push notifications

jaclync opened the pull request:
Fixes #5032 (the last part)
Project thread p91TBi-5Oe-p2

Just one reviewer is required for code review.

⚠️ Checklist before merging ⚠️ 
- [x] Backend Differential is merged and deployed with the same target app version
- [x] Revert app version change to 8.0 https://github.com/woocommerce/woocommerce-ios/pull/5299/commits/cad369bbe8242368fd4d9b8437be3143bd4e4303
- [x] Update release notes

## Why

To support the new notification content for multi-store push notifications, we made all the notification payload changes in server side (D68195). Since the payload format also changes, we need to update how we parse the payload when handling the notification tap action. The new payload format also includes a new field `subtitle`, and this PR updated the in-app notice UI to show the subtitle.

### New payload format

Currently, we only have one body text `alert` in the payload:

```json
{
    ...
    "aps": {
        "alert": "kayaipsum left a review on Album",
        ...
    }
}
```

The new payload replaces the `alert` value from a string with a dictionary with `title`, `subtitle`, and `body` (the [full list of `alert` keys](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification#2943365)). Additionally, it has a [thread ID](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification#2943363) that groups notifications by type and site (no client side changes required):

```json
{
    ...
    "aps": {
        "alert": {
            "title": "kayaipsum left a review on Album",
            "subtitle": "{{site name for store review}}", // no subtitle for store order notification because the site name is already shown in the body text
            "body": "Very nice!",
        },
        ...
    },
    "thread-id": "{{site ID}} - {{store_review / store_order}}" // the thread ID groups notifications by type and site
}
```

## Changes

- Turned on `pushNotificationsForAllStores` feature flag
- Added optional `subtitle` to `PushNotification` struct, and parsed `PushNotification` based on the feature flag
- Added subtitle to in-app notice and its presenter when the app receives a review notification in the foreground

## Testing

Because the backend changes are still pending review and not deployed yet, please test it with dev MC for now. I'm hoping for the feature to launch in release 7.9.

Prerequisites:
To test it with dev MC, please refer to the steps in p91TBi-6eM-p2#testing-push-notifications
Dev MC only works for a8c WP.com accounts, please make sure your a8c WP.com account is connected to more than one store.

### Receiving push notifications in the background

- In your dev sandbox, check out D68195 (`arc patch D68195`)
- Build the app onto a device
- Log in to your a8c WP.com account
- Put the app to the background or close the app
- Place two orders on one of your sites --> the app should receive two push notifications that say `You have a new order! 🎉`  (behavior in production)
- In Dev MC push notifications debug console (link in p91TBi-6eM-p2#testing-push-notifications), resend the last 2 store order notifications --> the app should receive two push notifications that say `You have a new order! 🎉` in the title, and `New order for {{order amount}} on {{site name}}` in the body. These 2 notifications should be grouped in Notification Center (they might not be grouped immediately, you can check back in a minute)
- Leave a review on two different products on one of your sites --> the app should receive two push notifications that say `{{customer name}} left a review on {{product name}}` (behavior in production)
- In Dev MC push notifications debug console, resend the last 2 store review notifications --> the app should receive two push notifications that say `You have a new review! ⭐️` in the title, store name in the subtitle, and `{{customer name}} left a review on {{product name}}: {{review content}}` in the body. These 2 notifications should be grouped in Notification Center (they might not be grouped immediately, you can check back in a minute)
- Feel free to try resending other store order/review notifications from different stores in Dev MC

### Receiving push notifications in the foreground

- Open the app and leave it in the foreground
- In Dev MC push notifications debug console, resend any store order notification --> the app should show an in-app notice that says `You have a new order! 🎉` in the title, and `New order for {{order amount}} on {{site name}}` in the body
- In Dev MC push notifications debug console, resend any store review notification --> the app should show an in-app notice that says `You have a new review! ⭐️` in the title, store name in the subtitle, and `{{customer name}} left a review on {{product name}}: {{review content}}` in the body

### Tapping on push notifications

- In the app, go to Settings > Switch Store. Switch to a store that is not related to the push notifications you resent from Dev MC above
- Tap on a store order notification --> the app should switch to the store of the notification, and display order details. an in-app notice should show about the store switch "Switched to {{store name}}"
- Tap on a store review notification --> the app should display review details

### Regression (for release 7.8 and if this feature does not ship in time for 7.9)

Feel free to skip this part. @jaclync will make sure to test this throughout the review process.

- Update app version to 7.8 if D68195 has been deployed
- In Xcode, in `DefaultFeatureFlagService`, return `true` for `pushNotificationsForAllStores` feature flag
- Place an order on the selected site --> the app should receive a push notification that says `You have a new order! 🎉`
- Place an order on a different site --> the app should not receive a push notification
- Leave a review on a product on the selected site --> the app should receive a push notification that says `{{customer name}} left a review on {{product name}}`. This push notification should be in the same group as the store order notification above
- Tap on the store order notification --> the app should show the order details
- Tap on the store review notification --> the app should show the review details

## Example screenshots

Notification Center | in-app notice for review notification
-- | --
![image](https://user-images.githubusercontent.com/1945542/139212243-b001bd65-2814-4546-a686-2d4efe61900b.png) | ![IMG_3371](https://user-images.githubusercontent.com/1945542/139212286-c185950c-573f-4e02-bf90-b6067b9455c1.jpg)



Update release notes:
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
Fix typos in `DevicesRemote`.

jaclync committed:
Update `PushNotification` to parse alert title and optional message from new push notification payload format.

jaclync committed:
Update/add unit tests for updated notification payload when multi-store push notifications are enabled.

jaclync committed:
Add a test case on inactive notification when multi-store push notifications are disabled.

jaclync committed:
Use version 7.8.0.0 for testing push notifications.

jaclync committed:
Bump version to 7.9.0.0 since we are not releasing multi-store push notifications this upcoming release.

jaclync committed:
NoticeView: add a subtitle label for optional alert subtitle.

jaclync committed:
Fix unit tests.

jaclync committed:
Update unit tests to test on payload subtitle.

jaclync committed:
Merge branch 'develop' into issue/5032-new-push-notification-payload

jaclync committed:
Merge branch 'develop' into issue/5032-new-push-notification-payload

# Conflicts:
#	config/Version.Public.xcconfig

jaclync committed:
Merge branch 'develop' into issue/5032-new-push-notification-payload

# Conflicts:
#	config/Version.Public.xcconfig

jaclync committed:
Turn on push notifications for all stores feature flag.

shiki commented:
@jaclync My WordPress.com sandbox was deleted a while ago. I requested a new one pMz3w-e4d-p2. It'll be a while for me to review this since I'll be AFK for a couple of days. Please re-assign if this will be a problem!

jaclync commented:
Thanks for the heads up @shiki! No worries, I'll reassign the reviewer 😄 

Trying @pmusolino or @ealeksandrov if you have a dev sandbox, and have some time Friday or Monday to review this 😃 we're hoping to get this in for release 7.9, but the status of the backend PR is still unknown so we might also just wait for another release cycle. @astralbodies might be able to help with testing as well!

jaclync commented:
We decided to move this feature to release 8.0, so take your time reviewing 🙂

jaclync committed:
Update app version to 8.0.0.0 for testing.
Merge branch 'develop' into issue/5032-new-push-notification-payload

# Conflicts:
#	config/Version.Public.xcconfig

jaclync committed:
Merge branch 'develop' into issue/5032-new-push-notification-payload

shiki commented:
> In Dev MC push notifications debug console (link in p91TBi-6eM-p2#testing-push-notifications), resend the last 2 store order notifications

I'm not sure if I'm testing this right. 😐 I only see one push notification about the order in dev-mc

![image](https://user-images.githubusercontent.com/198826/139945413-53a0ec8d-ca78-4ea2-aa59-a31f2af88252.png)

And I think that's expected. So, when I try to resend the push notifications of the last 2 orders, I get the push notifications for the same order only: 

<img src="https://user-images.githubusercontent.com/198826/139945504-a36e4ce6-2fdd-4431-95a0-a0084dfc583d.png" width="300">

Also, they don't seem to be grouped together. What could I be missing? 

---

@pmusolino or @ealeksandrov feel free to still review this even though I'm commenting. I have a feeling my problem is because I'm a noob. 😅

shiki commented:
> These 2 notifications should be grouped in Notification Center (they might not be grouped immediately, you can check back in a minute)

Update. 🤣 So it looks like the notifications will be grouped if I received the push notifications while the device is locked. 

<img src="https://user-images.githubusercontent.com/198826/139954690-9356cc53-9d32-48f3-95ea-da2d79ca13c8.png" width="300">

shiki commented on WooCommerce/Classes/Notifications/ApplicationAdapter.swift:
```diff
@@ -40,8 +40,8 @@ extension UIApplication: ApplicationAdapter {
     ///
-    func presentInAppNotification(message: String) {
-        let notice = Notice(title: message, message: nil, feedbackType: .success)
+    func presentInAppNotification(title: String, subtitle: String?, message: String?) {
```
We never handled tapping on the Notices, did we? I can receive a notice but when I tap on it, nothing happens. 🤷  Just wanted to confirm.

jaclync replied:
That's correct 😅 I created an issue about this, and I had some other thoughts while writing the issue https://github.com/woocommerce/woocommerce-ios/issues/5330 - would like to hear what you and the team think 🙂

shiki commented on WooCommerce/Classes/Notifications/PushNotificationsManager.swift:
```diff
@@ -519,15 +531,27 @@ private extension PushNotificationsManager {
-              let type = userInfo.string(forKey: APNSKey.type),
-              let noteKind = Note.Kind(rawValue: type) else {
-            return nil
+    static func from(userInfo: [AnyHashable: Any], pushNotificationsForAllStoresEnabled: Bool) -> PushNotification? {
```
Maybe a more apt argument name is `usingV2PayloadFormat`. 

```suggestion
    static func from(userInfo: [AnyHashable: Any], usingV2PayloadFormat: Bool) -> PushNotification? {
```

🤷

jaclync replied:
This code is to be deleted after the multi-store push notifications release is stable, I thought it might be more obvious with the feature flag name? 😄

shiki commented on WooCommerce/Classes/Tools/Notices/Notice.swift:
```diff
@@ -11,6 +11,10 @@ struct Notice {
     ///
     let title: String
 
+    /// An optional subtitle for the notice
+    ///
+    let subtitle: String?
+
     /// An optional subtitle for the notice
     ///
     let message: String?
```
ℹ️ We might want to consider clarifying the difference between `subtitle` and `message`. Both have the same descriptions. Maybe a description indicating how they are displayed.

jaclync replied:
Great suggestion! Updated in https://github.com/woocommerce/woocommerce-ios/pull/5299/commits/eea115f296e1ab8e5f8203d12e2c479518542469 based on the [Apple documentation](https://developer.apple.com/documentation/usernotifications/unmutablenotificationcontent/1649873-subtitle) on these fields

shiki commented on WooCommerce/WooCommerceTests/Notifications/PushNotificationsManagerTests.swift:
```diff
@@ -346,10 +475,55 @@ final class PushNotificationsManagerTests: XCTestCase {
+        // Then
+        XCTAssertEqual(emittedNotifications.count, 1)
+
+        let emittedNotification = emittedNotifications.first!
```
Maybe `XCTUnwrap` would be better. 🙂 

```suggestion
        let emittedNotification = try XCTUnwrap(emittedNotifications.first)
```

jaclync replied:
Nice catch! Updated in https://github.com/woocommerce/woocommerce-ios/pull/5299/commits/eea115f296e1ab8e5f8203d12e2c479518542469

shiki commented on WooCommerce/WooCommerceTests/Notifications/PushNotificationsManagerTests.swift:
```diff
@@ -346,10 +475,55 @@ final class PushNotificationsManagerTests: XCTestCase {
     }
 
-    func testItDoesNotEmitForegroundNotificationsWhenItReceivesANotificationWhileAppIsNotActive() {
+    func test_it_does_not_emit_foreground_notifications_when_it_receives_a_notification_while_app_is_not_active() {
```
Thanks for updating the function names!

shiki reviewed (approved):
Looks good to me! I only have minor comments that you can ignore. 😁 

### Testing

- [x] Receiving push notifications in the background
    - I had trouble with the grouping but you're right, it takes a while for the notifications to be grouped by iOS.
- [x] Receiving push notifications in the foreground
    - It looks like we never handled tapping on the notifications. I wonder if we should log that for the future.
- [x] Tapping on push notifications

jaclync committed:
Merge branch 'develop' into issue/5032-new-push-notification-payload

# Conflicts:
#	config/Version.Public.xcconfig

jaclync committed:
Updates based on CR suggestions.

jaclync commented:
Thanks a lot for testing with dev MC after requesting sandbox @shiki 🙏 Updated the PR and responded to your comments if you'd like another look!

> I only see one push notification about the order in dev-mc

Ah that's correct, each new order notification replaces the previous one for the same store. We can still send the same order notification multiple times from dev MC.

pmusolino commented:
> @pmusolino or @ealeksandrov feel free to still review this even though I'm commenting. I have a feeling my problem is because I'm a noob. 😅

I'm struggling for the second time in a month with our developer provisioning profiles 😰

shiki reviewed (approved):
:shipit:

jaclync committed:
Merge branch 'develop' into issue/5032-new-push-notification-payload

jaclync committed:
Merge branch 'develop' into issue/5032-new-push-notification-payload

# Conflicts:
#	Experiments/Experiments/DefaultFeatureFlagService.swift

jaclync committed:
Revert app version change, this will be updated after code freeze.

jaclync committed:
Update release notes.

jaclync commented:
The backend changes D68195 were deployed, and I tested the PR branch with the app version set to 8.0.0.0 and also 7.9.0.0 with the feature flag off. Merging this PR now!

astralbodies commented:
@jaclync 👏

cuemarie commented:
Hey there! Tested what I could today without setting up DevMC - not sure if that impacts what's noted below. 

I was able to experience all features except the following:

> **Tapping on push notifications**
    In the app, go to Settings > Switch Store. Switch to a store that is not related to the push notifications you resent from Dev MC above
    Tap on a store order notification --> the app should switch to the store of the notification, and display order details. an in-app notice should show about the store switch "Switched to {{store name}}"
    Tap on a store review notification --> the app should display review details

**App Open to STORE1, notification for STORE1:**
- tapping on a new review or new order cleared the in-app notification, but did not take me tot he orders or reviews page, and did not show any more details

**App Open to STORE2, notification for STORE1:**
 - same results as above; tapping on the in-app notification only clears it; did not switch stores

iPhone XR - iOS 14.8.1, WooCommerce App 8.0
